drivers/bluetooth/hci/rpmsg_nrf53.c

```diff
@@ -0,0 +1,241 @@
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME bt_hci_driver_nrf53
+#include "common/log.h"
+
+void bt_rpmsg_rx(u8_t *data, size_t len);
+
+static K_SEM_DEFINE(sync_sem, 0, 1);
+
+static struct device *ipm_tx_handle;
+static struct device *ipm_rx_handle;
+
+/* Configuration defines */
+
+#define SHM_START_ADDR      (DT_IPC_SHM_BASE_ADDRESS + 0x400)
+#define SHM_SIZE            0x7c00
+#define SHM_DEVICE_NAME     "sram0.shm"
+
+#define VRING_COUNT         2
+#define VRING_TX_ADDRESS    (SHM_START_ADDR + SHM_SIZE - 0x400)
+#define VRING_RX_ADDRESS    (VRING_TX_ADDRESS - 0x400)
+#define VRING_ALIGNMENT     4
+#define VRING_SIZE          16
+
+#define VDEV_STATUS_ADDR    DT_IPC_SHM_BASE_ADDRESS
+
+/* End of configuration defines */
+
+static metal_phys_addr_t shm_physmap[] = { SHM_START_ADDR };
+static struct metal_device shm_device = {
+	.name = SHM_DEVICE_NAME,
+	.bus = NULL,
+	.num_regions = 1,
+	.regions = {
+		{
+			.virt       = (void *) SHM_START_ADDR,
+			.physmap    = shm_physmap,
+			.size       = SHM_SIZE,
+			.page_shift = 0xffffffff,
+			.page_mask  = 0xffffffff,
+			.mem_flags  = 0,
+			.ops        = { NULL },
+		},
+	},
+	.node = { NULL },
+	.irq_num = 0,
+	.irq_info = NULL
+};
+
+static struct virtqueue *vq[2];
+static struct rpmsg_endpoint ep;
+
+static unsigned char virtio_get_status(struct virtio_device *vdev)
+{
+	return VIRTIO_CONFIG_STATUS_DRIVER_OK;
+}
+
+static void virtio_set_status(struct virtio_device *vdev, unsigned char status)
+{
+	sys_write8(status, VDEV_STATUS_ADDR);
+}
+
+static u32_t virtio_get_features(struct virtio_device *vdev)
+{
+	return BIT(VIRTIO_RPMSG_F_NS);
+}
+
+static void virtio_set_features(struct virtio_device *vdev, u32_t features)
+{
+	/* No need for implementation */
+}
+
+static void virtio_notify(struct virtqueue *vq)
+{
+	int status;
+
+	status = ipm_send(ipm_tx_handle, 0, 0, NULL, 0);
+	if (status != 0) {
+		BT_ERR("ipm_send failed to notify: %d", status);
+	}
+}
+
+const struct virtio_dispatch dispatch = {
+	.get_status = virtio_get_status,
+	.set_status = virtio_set_status,
+	.get_features = virtio_get_features,
+	.set_features = virtio_set_features,
+	.notify = virtio_notify,
+};
+
+static void ipm_callback(void *context, u32_t id, volatile void *data)
+{
+	BT_DBG("Got callback of id %u", id);
+	virtqueue_notification(vq[0]);
+}
+
+int endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
+	u32_t src, void *priv)
+{
+	BT_DBG("Received message of %u bytes.", len);
+	BT_HEXDUMP_DBG((uint8_t *)data, len, "Data:");
+
+	bt_rpmsg_rx(data, len);
+
+	return RPMSG_SUCCESS;
+}
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+void ns_bind_cb(struct rpmsg_device *rdev, const char *name, u32_t dest)
+{
+	(void)rpmsg_create_ept(&ep,
+				rdev,
+				name,
+				RPMSG_ADDR_ANY,
+				dest,
+				endpoint_cb,
+				rpmsg_service_unbind);
+
+	k_sem_give(&sync_sem);
+}
+
+int bt_rpmsg_platform_init(void)
+{
```

Comment:
This implementation is working but is not optimized as use one OpenAMP instance for one RPMSg service.
We should be able to share these vrings with some other rpmsg client drivers (refer to https://github.com/zephyrproject-rtos/zephyr/issues/20419 discussion).
 Can probably be addressed later in a separate thread...

Reply:
Yes, I think so @arnopo , let's take this on separate thread :)